File: androidApp/src/main/java/com/winthan/watchmekmp/android/home/HomeViewModel.kt
package com.winthan.watchmekmp.android.home

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.winthan.watchmekmp.domain.model.Movie
import com.winthan.watchmekmp.domain.usecase.HomeScreenUseCase
import kotlinx.coroutines.launch

class HomeViewModel(
    private val homeScreenUseCase: HomeScreenUseCase
) : ViewModel() {

    var uiState by mutableStateOf(HomeScreenState())
        private set

    private var currentPage = 1

    init {
        loadMovies()
    }

    private fun loadMovies() {
        loadMovieType(MovieType.POPULAR)
        loadMovieType(MovieType.NOW_PLAYING)
        loadMovieType(MovieType.TOP_RATED)
        loadMovieType(MovieType.UPCOMING)
    }

    private fun loadMovieType(movieType: MovieType) {
        viewModelScope.launch {
            updateUiStateLoading(movieType, true)
            try {
                val result = when (movieType) {
                    MovieType.POPULAR -> homeScreenUseCase.getPopularMovieUseCase(currentPage)
                    MovieType.NOW_PLAYING -> homeScreenUseCase.getNowPlayingMovieUseCase(currentPage)
                    MovieType.TOP_RATED -> homeScreenUseCase.getTopRatedMovieUseCase(currentPage)
                    MovieType.UPCOMING -> homeScreenUseCase.getUpcomingMovieUseCase(currentPage)
                }
                updateUiStateSuccess(movieType, result)
            } catch (e: Exception) {
                updateUiStateError(movieType, e.localizedMessage ?: "Unknown error")
            }
        }
    }

    private fun updateUiStateLoading(movieType: MovieType, isLoading: Boolean) {
        uiState = when (movieType) {
            MovieType.POPULAR -> uiState.copy(isPopularMovieLoading = isLoading)
            MovieType.NOW_PLAYING -> uiState.copy(isNowPlayingMovieLoading = isLoading)
            MovieType.TOP_RATED -> uiState.copy(isTopRatedMovieLoading = isLoading)
            MovieType.UPCOMING -> uiState.copy(isUpcomingMovieLoading = isLoading)
        }
    }

    private fun updateUiStateSuccess(movieType: MovieType, movies: List<Movie>) {
        uiState = when (movieType) {
            MovieType.POPULAR -> uiState.copy(isPopularMovieLoading = false, popularMovies = movies)
            MovieType.NOW_PLAYING -> uiState.copy(isNowPlayingMovieLoading = false, nowPlayingMovies = movies)
            MovieType.TOP_RATED -> uiState.copy(isTopRatedMovieLoading = false, topRatedMovies = movies)
            MovieType.UPCOMING -> uiState.copy(isUpcomingMovieLoading = false, upcomingMovies = movies)
        }
    }

    private fun updateUiStateError(movieType: MovieType, errorMessage: String) {
        uiState = when (movieType) {
            MovieType.POPULAR -> uiState.copy(isPopularMovieLoading = false, popularMovieErrorMessage = errorMessage)
            MovieType.NOW_PLAYING -> uiState.copy(isNowPlayingMovieLoading = false, nowPlayingMovieErrorMessage = errorMessage)
            MovieType.TOP_RATED -> uiState.copy(isTopRatedMovieLoading = false, topRatedMovieErrorMessage = errorMessage)
            MovieType.UPCOMING -> uiState.copy(isUpcomingMovieLoading = false, upcomingMovieErrorMessage = errorMessage)
        }
    }

    enum class MovieType {
        POPULAR, NOW_PLAYING, TOP_RATED, UPCOMING
    }


}

data class HomeScreenState(
    var isPopularMovieLoading: Boolean = false,
    var popularMovies: List<Movie> = emptyList(),
    var popularMovieErrorMessage: String? = null,
    var isNowPlayingMovieLoading: Boolean = false,
    var nowPlayingMovies: List<Movie> = emptyList(),
    var nowPlayingMovieErrorMessage: String? = null,
    var isTopRatedMovieLoading: Boolean = false,
    var topRatedMovies: List<Movie> = emptyList(),
    var topRatedMovieErrorMessage: String? = null,
    var isUpcomingMovieLoading: Boolean = false,
    var upcomingMovies: List<Movie> = emptyList(),
    var upcomingMovieErrorMessage: String? = null
)

File: iosApp/iosApp/home/HomeViewModel.swift
//
//  HomeViewModel2.swift
//  iosApp
//
//  Created by Win Than Htike on 21/2/2567 BE.
//  Copyright © 2567 BE orgName. All rights reserved.
//

import Foundation
import shared

enum MovieType {
    case popular, nowPlaying, topRated, upcoming
}

struct HomeScreenState {
    var isPopularMovieLoading = false
    var popularMovies: [Movie] = []
    var popularMovieErrorMessage: String? = nil

    var isNowPlayingMovieLoading = false
    var nowPlayingMovies: [Movie] = []
    var nowPlayingMovieErrorMessage: String? = nil

    var isTopRatedMovieLoading = false
    var topRatedMovies: [Movie] = []
    var topRatedMovieErrorMessage: String? = nil

    var isUpcomingMovieLoading = false
    var upcomingMovies: [Movie] = []
    var upcomingMovieErrorMessage: String? = nil
}


@MainActor
class HomeViewModel: ObservableObject {
    
    private var homeScreenUseCase: HomeScreenUseCase = HomeScreenUseCase.init(
        getUpcomingMovieUseCase: GetUpcomingMovieUseCase(),
        getTopRatedMovieUseCase: GetTopRatedMovieUseCase(),
        getNowPlayingMovieUseCase: GetNowPlayingMovieUseCase(),
        getPopularMovieUseCase: GetPopularMovieUseCase()
    )
    private var currentPage = 1

    @Published var uiState = HomeScreenState()

    func loadMovies() {
        loadMovieType(.popular)
        loadMovieType(.nowPlaying)
        loadMovieType(.topRated)
        loadMovieType(.upcoming)
    }

    private func loadMovieType(_ movieType: MovieType) {
        Task {
            do {
                let result: [Movie] = try await {
                    switch movieType {
                    case .popular:
                        return try await homeScreenUseCase.getPopularMovieUseCase.invoke(page: Int32(currentPage))
                    case .nowPlaying:
                        return try await homeScreenUseCase.getNowPlayingMovieUseCase.invoke(page: Int32(currentPage))
                    case .topRated:
                        return try await homeScreenUseCase.getTopRatedMovieUseCase.invoke(page: Int32(currentPage))
                    case .upcoming:
                        return try await homeScreenUseCase.getUpcomingMovieUseCase.invoke(page: Int32(currentPage))
                    }
                }()
                DispatchQueue.main.async {
                    self.updateUiStateSuccess(movieType, movies: result)
                }
            } catch {
                DispatchQueue.main.async {
                    self.updateUiStateError(movieType, errorMessage: error.localizedDescription)
                }
            }
        }
    }

    private func updateUiStateLoading(_ movieType: MovieType, isLoading: Bool) {
        switch movieType {
        case .popular:
            uiState.isPopularMovieLoading = true
        case .nowPlaying:
            uiState.isNowPlayingMovieLoading = true
        case .topRated:
            uiState.isTopRatedMovieLoading = true
        case .upcoming:
            uiState.isUpcomingMovieLoading = true
        }
    }

    private func updateUiStateSuccess(_ movieType: MovieType, movies: [Movie]) {
        switch movieType {
        case .popular:
            uiState.isPopularMovieLoading = false
            uiState.popularMovies = movies
        case .nowPlaying:
            uiState.isNowPlayingMovieLoading = false
            uiState.nowPlayingMovies = movies
        case .topRated:
            uiState.isTopRatedMovieLoading = false
            uiState.topRatedMovies = movies
        case .upcoming:
            uiState.isUpcomingMovieLoading = false
            uiState.upcomingMovies = movies
        }
    }

    private func updateUiStateError(_ movieType: MovieType, errorMessage: String) {
        switch movieType {
        case .popular:
            uiState.popularMovieErrorMessage = errorMessage
        case .nowPlaying:
            uiState.nowPlayingMovieErrorMessage = errorMessage
        case .topRated:
            uiState.topRatedMovieErrorMessage = errorMessage
        case .upcoming:
            uiState.upcomingMovieErrorMessage = errorMessage
        }
    }
}
